app/assets/scripts/components/upload-form.js

```
@@ -62,9 +62,8 @@ var UploadForm = React.createClass({
           failures = ['No data provided'];
           this.logOutput(failures);
         }
-        // Check header on line 1
+        // Check header on first line
         if (line === 0) this.checkHeader(data, failures);
```

You're passing the `failures` array around to be freely modified by other functions. You should try to avoid this as it's very difficult to keep track of the result of the function. The function should instead return values and they should be added to the main array.

---

If I understand correctly, I think this has been fixed. The logOutput function, which is now called setErrorState, now only receives the array of logs and sets the state in response. The metadata and failures are only appended to in the main csvreader stream, and aren't modified by any other functions until they are rendered.

---

the function `checkHeader` is still accepting an array as a parameter which is being modified internally, instead of returning a new array of failures.

Also the `checkHeader` should do just that, check the header for errors. `this.logOutput(failures);` should be done outside that function.

Perhaps something like:
```
if (line === 0) {
  failures = failures.concat(this.checkHeader(data));
}
if (failures.length) {
  this.logOutput(failures);
}
```